nflows/distributions/uniform.py

```diff
@@ -3,32 +3,61 @@
 import torch
 from torch import distributions
 
+from nflows.distributions.base import Distribution
+from nflows.utils import torchutils
 
-class BoxUniform(distributions.Independent):
+
+class BoxUniform(Distribution):
     def __init__(
         self,
         low: Union[torch.Tensor, float],
-        high: Union[torch.Tensor, float],
-        reinterpreted_batch_ndims: int = 1,
+        high: Union[torch.Tensor, float]    
     ):
         """Multidimensionqal uniform distribution defined on a box.
-        
-        A `Uniform` distribution initialized with e.g. a parameter vector low or high of length 3 will result in a /batch/ dimension of length 3. A log_prob evaluation will then output three numbers, one for each of the independent Uniforms in the batch. Instead, a `BoxUniform` initialized in the same way has three /event/ dimensions, and returns a scalar log_prob corresponding to whether the evaluated point is in the box defined by low and high or outside. 
-    
-        Refer to torch.distributions.Uniform and torch.distributions.Independent for further documentation.
-    
+            
         Args:
             low (Tensor or float): lower range (inclusive).
             high (Tensor or float): upper range (exclusive).
             reinterpreted_batch_ndims (int): the number of batch dims to
                                              reinterpret as event dims.
         """
-
-        super().__init__(
-            distributions.Uniform(low=low, high=high), reinterpreted_batch_ndims
-        )
-
-
+        super().__init__()
+        if low.shape != high.shape:
+            raise ValueError(
+                "low and high are not of the same size"
+            )
+
+        if not (low < high).byte().all():
+            raise ValueError(
+                "low has elements that are higher than high"
+            )
+
+        self._shape = low.shape
+        self._low = low
+        self._high = high
+        self._log_prob_value = -torch.sum(torch.log(high - low))
+
+    def _log_prob(self, inputs, context):
+        # Note: the context is ignored.
```

Comment:
Can we make the context set new lower/upper bounds INSIDE the current bounds? I.e. let the context select a subregion of the domain? context_low = low + (high-low)*sigmoid(context)?

Reply:
also really like this idea^. Maybe a ConditionalUniform class?

Reply:
Although if a point is outside the distribution, things will blow up. Maybe useful to have another distribution with a softer drop outside the uniform bounds?

Reply:
Yes, I believe too that the conditional distribution should be inside the unconditional one, and that's best done by picking a transform that guarantees it. I think one way to argue for having a conditional distribution that is a box inside the original box is via Bayes theorem for conditional distributions P(A|B) = P(A & B) / P(B). The P(A & B) is the intersection of the original box with a new box, which is a sub-box? But maybe it's not a very good concept to use in this case.

I also like your other idea of allowing flows to use base distributions that don't have a context. Another way to do that is by making the context argument optional in the flow's base class,  "_log_prob(self, inputs, context**=None**)". And then forward it like self._distribution.log_prob(noise) only if context is None. That way you can use use the distributions that don't have a context in flows, but only if the flow doesn't have a context either. This might prevent people from accidentally using a non-context distribution with a flow that does have a context?

Btw, do you know if we can use Splines instead of Afine in the conditional moons example? I tried it, but I get errors. It perhaps look like splines don't work with conditional Gaussian distributions as a base distribution? Do splines require distributions with a finite domain? If so, then having a conditional uniform distribution might make it possible?

Reply:
I did get the spline code working at some point, I don't remember if I made changes but I don't think you need to.

Re: the context, actually it should work with a context-less base distribution and context-full transformations! This is what I need this modification for actually.